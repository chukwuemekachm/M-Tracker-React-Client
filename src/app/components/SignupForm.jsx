import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import '../assets/css/auth.css';

/**
 * @description The Sign up form component
 *
 * @param {object} bindValues The user details to bind to the input fields
 * @param {function} handleSubmit The method to handle form submission
 * @param {function} handleChange The method to handle input changes
 *
 * @returns {object}
 */
const SignupForm = ({ bindValues, handleSubmit, handleChange }) => (
  <div className="row mb-4 auth-form">
    <div className="col-md-4" />
    <div className="col-md-4">
      <div className="card text-center">
        <h3 className="text-center mt-5">
          Sign up to use Maintenance Tracker
        </h3>
        <i className="icon ion-md-construct text-center ion-icon" />
        <div className="card-body mb-4 p-4">
          <form onSubmit={handleSubmit}>
            <div className="input-group mb-4">
              <input
                type="text"
                className="form-control text"
                placeholder="First name"
                name="firstname"
                onChange={handleChange}
                value={bindValues.firstname}
              />
            </div>
            <div className="input-group mb-4">
              <input
                type="text"
                className="form-control text"
                placeholder="Last name"
                name="lastname"
                onChange={handleChange}
                value={bindValues.lastname}
              />
            </div>
            <div className="input-group mb-4">
              <input
                type="email"
                className="form-control text"
                placeholder="Email"
                name="email"
                onChange={handleChange}
                value={bindValues.email}
              />
            </div>
            <div className="input-group mb-4">
              <input
                type="password"
                className="form-control text"
                placeholder="Password"
                name="password"
                onChange={handleChange}
                value={bindValues.password}
              />
            </div>
            <div className="input-group mb-4">
              <input
                type="password"
                className="form-control text"
                placeholder="Confirm password"
                name="confirmPassword"
                onChange={handleChange}
                value={bindValues.confirmPassword}
              />
            </div>
            <button
              type="submit"
              className="btn btn-success btn-md btn-block mb-3"
            >
              Signup
            </button>
            <span>
              Have an account?
              {' '}
              <Link to="/login">
                Login
              </Link>
            </span>
          </form>
        </div>
      </div>
    </div>
    <div className="col-md-4" />
  </div>
);

SignupForm.propTypes = {
  handleChange: PropTypes.func.isRequired,
  handleSubmit: PropTypes.func.isRequired,
  bindValues: PropTypes.shape({}).isRequired,
};

export default SignupForm;
