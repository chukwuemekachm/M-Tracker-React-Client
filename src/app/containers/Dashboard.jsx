import React, { Component } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import DefaultNavBar from '../components/NavBar';
import userRequestActions from '../actions/userRequestActions';
import UserRequest from '../components/UserRequest';
import '../assets/css/dashboard.css';
import SideNav from '../components/SideNav';
import CreateRequestForm from '../components/CreateRequest';
import filterRequests, { navigateFilter } from '../actions/filterRequestsAction';
import getAllRequests from '../actions/adminRequestActions';

/**
 * @description Dashboard component
 *
 * @class
 *
 * @extends React.Component
 */
export class Dashboard extends Component {
  /**
   * @constructor
   *
   * @param {object} props The injected props
   */
  constructor(props) {
    super(props);
    this.state = {};
  }

  /**
   * @description Evaluates the user's authentication
   * Calls the getRequest method to fetch the user's request
   *
   * @function
   */
  componentDidMount() {
    const {
      authenticated, history, getRequests, filter, admin, getAdminRequests,
    } = this.props;
    if (!authenticated) {
      return history.push('/login');
    }
    if (admin) {
      return getAdminRequests().then((response) => {
        if (response.code === 401) return history.push('/login');
        const { search } = history.location;
        return filter(search.substring(1));
      });
    }
    return getRequests().then((response) => {
      if (response.code === 401) return history.push('/login');
      const { search } = history.location;
      return filter(search.substring(1));
    });
  }

  /**
   * @description Renders the component on a dom node
   *
   * @function
   *
   * @returns {object} The components to render
   */
  render() {
    const {
      requests, history, createRequest, filteredRequests,
    } = this.props;
    return (
      <div>
        <div className="dashboard-body">
          <DefaultNavBar />
          <div className="container mt-5">
            <div className="row">
              <div className="col-md-3">
                <SideNav requests={requests} handleNavigation={navigateFilter} />
              </div>
              <div className="col-md-8">
                {
                  filteredRequests[0]
                    ? filteredRequests.map(request => (
                      <Link to={`/${request.id}`} key={request.id}>
                        <UserRequest key={request.id} {...request} />
                      </Link>
                    ))
                    : (
                      <div>
                        {
                          requests[0] && !filteredRequests[0]
                            ? (
                              <p className="text-center">
                                No requests exist for this category
                              </p>
                            )
                            : (
                              <p className="text-center">
                                You have no requests please click the
                                {' '}
                                create request button to make a new request.
                              </p>
                            )
                        }
                      </div>
                    )
                }
              </div>
            </div>
            <div className="col-md-1" />
          </div>
        </div>
        <CreateRequestForm history={history} create={createRequest} />
      </div>
    );
  }
}

Dashboard.propTypes = {
  requests: PropTypes.arrayOf(PropTypes.object).isRequired,
  authenticated: PropTypes.bool.isRequired,
  history: PropTypes.shape({}).isRequired,
  getRequests: PropTypes.func.isRequired,
  createRequest: PropTypes.func.isRequired,
  filter: PropTypes.func.isRequired,
  filteredRequests: PropTypes.arrayOf(PropTypes.object).isRequired,
  getAdminRequests: PropTypes.func.isRequired,
  admin: PropTypes.bool.isRequired,
};

export const mapStateToProps = (state, ownprops) => ({
  requests: state.requests.allRequests,
  filteredRequests: state.requests.filteredRequests,
  authenticated: state.auth.authenticated,
  history: ownprops.history,
  admin: state.auth.user.admin,
});

export const mapDispatchToProps = dispatch => ({
  getAdminRequests: () => dispatch(getAllRequests()),
  getRequests: () => dispatch(userRequestActions.getAllAsync()),
  createRequest: payload => dispatch(userRequestActions.createRequest(payload)),
  filter: payload => dispatch(filterRequests(payload)),
});

export default connect(mapStateToProps, mapDispatchToProps)(Dashboard);
