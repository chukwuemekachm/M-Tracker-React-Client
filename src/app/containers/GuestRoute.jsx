import React from 'react';
import { Route, Redirect } from 'react-router-dom';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

/**
 * @description Has the same API as the Route component
 * It verifies if the user is authenticated or not
 * If the user is logged in it redirects to the dashboard
 * Else it renders the component
 *
 * @param {object} component The component to be rendered
 * @param {boolean} authenticated The authentication state of the user
 * @param {array} rest Other props injected to the component
 *
 * @returns {object}
 * */
export const GuestRoute = ({ component: Component, authenticated, ...rest }) => (
  <Route
    {...rest}
    render={props => (
      !authenticated
        ? <Component {...props} />
        : <Redirect to="/dashboard" />
    )}
  />
);

GuestRoute.propTypes = {
  component: PropTypes.func.isRequired,
  authenticated: PropTypes.bool.isRequired,
};

const mapStateToProps = state => ({
  authenticated: state.auth.authenticated,
});

export default connect(mapStateToProps, null)(GuestRoute);
