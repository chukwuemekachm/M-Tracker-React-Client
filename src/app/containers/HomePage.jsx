import React from 'react';
import { Link } from 'react-router-dom';
import DefaultNavBar from '../components/NavBar';
import Icon from '../components/Icon';
import Footer from '../components/Footer';
import '../assets/css/home.css';

/**
 * @description The Home page component
 *
 * @returns {object}
 */
const HomePage = () => (
  <div>
    <DefaultNavBar />
    <div className="jumbotron">
      <div className="banner">
        <div className="w3-display-middle">
          <h3 className="text-center">
            Facilities don
            {'\''}
            t need to go bad any longer,
            {' '}
            request a repair or maintenance schedule from us today.
          </h3>
          <br />
          <div className="row">
            <div className="col-md-3" />
            <div className="col-md-6">
              <Link to="/signup">
                <button type="button" className="btn btn-success btn-lg btn-block">
                  Get Started
                </button>
              </Link>
            </div>
            <div className="col-md-3" />
          </div>
        </div>
      </div>
    </div>
    <div className="container">
      <br />
      <h3 className="display-6 text-center">
        About Maintenance Tracker
      </h3>
      <br />
      <div className="container">
        <p className="text-center">
          Maintenance Tracker is an platform designed, to provide users with the oppurtunity of
          {' '}
          reaching out to operations or repairs department regarding repair or maintenance
          {' '}
          requests. It also enables users to monitor the status of their requests.
        </p>
      </div>
    </div>
    <br />
    <br />
    <div className="container">
      <h3 className="display-6 text-center">
        How it Works
      </h3>
      <br />
      <br />
      <div className="row w3-display-center">
        <Icon
          title="Authenticate"
          icon="https://res.cloudinary.com/dvcaeuvee/image/upload/v1538265100/undraw_login_jdch.svg"
        />
        <Icon
          title="Log a request"
          icon="https://res.cloudinary.com/dvcaeuvee/image/upload/v1538265099/undraw_add_file2_gvbb.svg"
        />
        <Icon
          title="Request gets approved"
          icon="https://res.cloudinary.com/dvcaeuvee/image/upload/v1538265100/undraw_hiring_cyhs.svg"
        />
        <Icon
          title="Request gets resolved"
          icon="https://res.cloudinary.com/dvcaeuvee/image/upload/v1538265100/undraw_confirmation_2uy0.svg"
        />
      </div>
    </div>
    <Footer />
  </div>
);

export default HomePage;
