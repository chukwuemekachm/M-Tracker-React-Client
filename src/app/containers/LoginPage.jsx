import React, { Component } from 'react';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import DefaultNavBar from '../components/NavBar';
import LoginForm from '../components/LoginForm';
import authActions from '../actions/authActions';
import Loader from '../components/Loader';

export class LoginPage extends Component {
  constructor(props) {
    super(props);
    this.state = {
      email: '',
      password: '',
    };
    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
  }

  handleChange(event) {
    this.setState({
      [event.target.name]: event.target.value,
    });
  }

  handleSubmit(event) {
    event.preventDefault();
    const { userLogin, history } = this.props;
    userLogin(this.state, history, 'login');
  }

  render() {
    const { loading } = this.props;
    return (
      <div>
        <DefaultNavBar />
        <LoginForm
          handleChange={this.handleChange}
          handleSubmit={this.handleSubmit}
          bindValues={this.state}
        />
        {
          loading
            ? <Loader />
            : ''
        }
      </div>
    );
  }
}

LoginPage.propTypes = {
  // message: PropTypes.string.isRequired,
  loading: PropTypes.bool.isRequired,
  userLogin: PropTypes.func.isRequired,
  history: PropTypes.shape({}).isRequired,
};

const mapStateToProps = (state, ownProps) => ({
  error: state.common.error,
  message: state.common.message,
  loading: state.common.loading,
  history: ownProps.history,
});

const mapDispatchToProps = dispatch => bindActionCreators({
  userLogin: (user, history, route) => (authActions.authAsync(user, history, route)),
}, dispatch);

export default connect(mapStateToProps, mapDispatchToProps)(LoginPage);
