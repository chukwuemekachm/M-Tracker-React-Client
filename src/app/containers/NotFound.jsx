import React from 'react';
import DefaultNavBar from '../components/NavBar';

/**
 * @description The Not found component
 *
 * @returns {object}
 */
const NotFound = () => (
  <div className="ch-aside">
    <DefaultNavBar />
    <div
      className="ch-row"
      style={{
        backgroundColor: 'rgba(0, 0, 0, 0.7)', color: 'white', width: '100%', minHeight: '100%', position: 'fixed',
      }}
    >
      <div className="ch-col-6">
        <h1 style={{ fontSize: '8em' }}>
          404
        </h1>
        <p>
          Oops, the page you are looking for was not found.
        </p>
      </div>
    </div>
  </div>
);

export default NotFound;
