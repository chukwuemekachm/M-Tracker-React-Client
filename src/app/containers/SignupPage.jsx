import React, { Component } from 'react';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import toastr from 'toastr';
import DefaultNavBar from '../components/NavBar';
import SignupForm from '../components/SignupForm';
import authActions from '../actions/authActions';
import { userError } from '../actions/commonActions';
import Loader from '../components/Loader';

export class SignupPage extends Component {
  constructor(props) {
    super(props);
    this.state = {
      firstname: '',
      lastname: '',
      email: '',
      password: '',
      confirmPassword: '',
    };
    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
  }

  handleChange(event) {
    this.setState({
      [event.target.name]: event.target.value,
    });
  }

  handleSubmit(event) {
    event.preventDefault();
    const { password, confirmPassword } = this.state;
    const { userSignup, userErrorHandler, history } = this.props;
    if (confirmPassword !== password) {
      userErrorHandler('Confirm Password must match the Password');
      toastr.error('Confirm Password must match the Password');
      return;
    }
    userSignup(this.state, history, 'signup');
  }

  render() {
    const { loading } = this.props;
    return (
      <div>
        <DefaultNavBar />
        <SignupForm
          handleChange={this.handleChange}
          handleSubmit={this.handleSubmit}
          bindValues={this.state}
        />
        {
          loading
            ? <Loader />
            : ''
        }
      </div>
    );
  }
}

SignupPage.propTypes = {
  loading: PropTypes.bool.isRequired,
  userErrorHandler: PropTypes.func.isRequired,
  userSignup: PropTypes.func.isRequired,
  history: PropTypes.shape({}).isRequired,
};

const mapStateToProps = (state, ownProps) => ({
  error: state.common.error,
  message: state.common.message,
  loading: state.common.loading,
  history: ownProps.history,
});

const mapDispatchToProps = dispatch => bindActionCreators({
  userSignup: (user, history, route) => (authActions.authAsync(user, history, route)),
  userErrorHandler: message => (userError(message)),
}, dispatch);

export default connect(mapStateToProps, mapDispatchToProps)(SignupPage);
