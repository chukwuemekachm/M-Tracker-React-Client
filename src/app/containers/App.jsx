import React from 'react';
import HomePage from './HomePage';
import '../assets/css/style.css';

const App = () => (
  <div>
    <HomePage />
  </div>
);

export default App;
