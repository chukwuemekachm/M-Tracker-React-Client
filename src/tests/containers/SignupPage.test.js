import React from 'react';
import { shallow } from 'enzyme';
import { SignupPage } from '../../app/containers/SignupPage';
import SignupForm from '../../app/components/SignupForm';

const userErrorHandler = jest.fn(() => true);
const userSignup = jest.fn();
const history = {};
const mockSubmitEvent = {
  preventDefault: jest.fn(),
};
const user = {
  firstname: 'John',
  lastname: 'Doe',
  email: 'user@test.com',
  password: 'Password1',
  confirmPassword: 'Password1',
};
const mockChangeEvent = (name, value) => ({
  target: {
    name,
    value,
  },
});

describe('SignupPage container', () => {
  it('should render without errors when props are falsy', () => {
    const wrapper = shallow(<SignupPage
      loading={false}
      error={false}
      message="test"
      userErrorHandler={userErrorHandler}
      userSignup={userSignup}
      history={history}
    />);
    expect(wrapper.find(SignupForm)).toHaveLength(1);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render without errors when props are truthy', () => {
    const wrapper = shallow(<SignupPage
      loading
      error
      message="test"
      userErrorHandler={userErrorHandler}
      userSignup={userSignup}
      history={history}
    />);
    expect(wrapper.find(SignupForm)).toHaveLength(1);
  });

  it('should handle password mismatch', () => {
    const wrapper = shallow(<SignupPage
      loading
      error
      message="test"
      userErrorHandler={userErrorHandler}
      userSignup={userSignup}
      history={history}
    />);
    const instance = wrapper.instance();
    instance.state.password = user.password;
    instance.state.confirmPassword = 'Mismatch1';
    instance.handleSubmit(mockSubmitEvent);
    expect(instance.handleSubmit(mockSubmitEvent)).toBeFalsy();
  });

  it('should handle form element changes', () => {
    const wrapper = shallow(<SignupPage
      loading
      error
      message="test"
      userErrorHandler={userErrorHandler}
      userSignup={userSignup}
      history={history}
    />);

    const instance = wrapper.instance();
    instance.handleChange(mockChangeEvent('email', user.email));
    instance.handleChange(mockChangeEvent('firstname', user.firstname));
    instance.handleChange(mockChangeEvent('lastname', user.lastname));
    instance.handleChange(mockChangeEvent('password', user.password));
    instance.handleChange(mockChangeEvent('confirmPassword', user.confirmPassword));

    expect(instance.state.email).toEqual(user.email);
    expect(instance.state.firstname).toEqual(user.firstname);
    expect(instance.state.lastname).toEqual(user.lastname);
    expect(instance.state.password).toEqual(user.password);
    expect(instance.state.confirmPassword).toEqual(user.confirmPassword);
  });

  it('should handle form submission', () => {
    const wrapper = shallow(<SignupPage
      loading
      error
      message="test"
      userErrorHandler={userErrorHandler}
      userSignup={userSignup}
      history={history}
    />);
    const instance = wrapper.instance();
    instance.state = user;
    expect(instance.state).toEqual(user);
    expect(instance.handleSubmit(mockSubmitEvent)).toBeTruthy();
  });
});
