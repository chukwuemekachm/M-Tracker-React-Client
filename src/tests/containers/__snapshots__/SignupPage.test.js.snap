// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignupPage container should render without errors when props are falsy 1`] = `
<div>
  <Connect(NavBar) />
  <SignupForm
    bindValues={
      Object {
        "confirmPassword": "",
        "email": "",
        "firstname": "",
        "lastname": "",
        "password": "",
      }
    }
    handleChange={[Function]}
    handleSubmit={[Function]}
  />
</div>
`;
